File: app/cache/redis_cache.py
"""
Centralized Redis cache service for workflow automation.

Provides a thread-safe, singleton Redis client with TTL-based caching
for frequently-read data like workflow definitions.
"""
import json
import logging
from typing import Any

import redis

from app.core.config import settings

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

_redis_client: redis.Redis | None = None


def get_redis_client() -> redis.Redis:
    """
    Get or create the singleton Redis client.

    Returns:
        redis.Redis: A connected Redis client instance.
    """
    global _redis_client
    if _redis_client is None:
        _redis_client = redis.Redis(
            host=settings.REDIS_HOST,
            port=int(settings.REDIS_PORT),
            db=0,
            decode_responses=True,
            socket_connect_timeout=5,
            socket_timeout=5,
            retry_on_timeout=True,
        )
        logger.info(f"Redis client created: {settings.REDIS_HOST}:{settings.REDIS_PORT}")
    return _redis_client


def cache_get(key: str) -> dict | None:
    """
    Get a value from the cache.

    Args:
        key: The cache key.

    Returns:
        The cached value as a dict, or None if not found.
    """
    try:
        client = get_redis_client()
        value = client.get(key)
        if value:
            return json.loads(value)
    except Exception as e:
        logger.warning(f"Redis cache_get error for key={key}: {e}")
    return None


def cache_set(key: str, value: Any, ttl: int = 60) -> None:
    """
    Set a value in the cache with a TTL.

    Args:
        key: The cache key.
        value: The value to cache (must be JSON-serializable).
        ttl: Time-to-live in seconds (default: 60).
    """
    try:
        client = get_redis_client()
        client.setex(key, ttl, json.dumps(value, default=str))
    except Exception as e:
        logger.warning(f"Redis cache_set error for key={key}: {e}")


def cache_delete(key: str) -> None:
    """
    Delete a value from the cache.

    Args:
        key: The cache key to delete.
    """
    try:
        client = get_redis_client()
        client.delete(key)
    except Exception as e:
        logger.warning(f"Redis cache_delete error for key={key}: {e}")


def cache_delete_pattern(pattern: str) -> None:
    """
    Delete all keys matching a pattern.

    Args:
        pattern: The glob pattern to match (e.g., 'workflow:*').
    """
    try:
        client = get_redis_client()
        keys = client.keys(pattern)
        if keys:
            client.delete(*keys)
    except Exception as e:
        logger.warning(f"Redis cache_delete_pattern error for pattern={pattern}: {e}")

File: rust-app/src/cache/redis_cache.rs
//! Centralized Redis cache service for workflow automation.
//!
//! Provides async Redis operations with TTL-based caching
//! for frequently-read data like workflow definitions.

use redis::AsyncCommands;
use tracing::warn;

/// Redis cache client wrapper.
pub struct RedisCache {
    client: redis::Client,
}

impl RedisCache {
    /// Create a new RedisCache from a Redis URL (e.g., "redis://redis:6379").
    pub fn new(redis_url: &str) -> Result<Self, redis::RedisError> {
        let client = redis::Client::open(redis_url)?;
        Ok(Self { client })
    }

    /// Get a value from the cache.
    ///
    /// Returns None if the key doesn't exist or on error (fail-open).
    pub async fn get(&self, key: &str) -> Option<String> {
        match self.client.get_multiplexed_async_connection().await {
            Ok(mut conn) => match conn.get::<_, Option<String>>(key).await {
                Ok(val) => val,
                Err(e) => {
                    warn!("Redis cache_get error for key={}: {}", key, e);
                    None
                }
            },
            Err(e) => {
                warn!("Redis connection error in cache_get: {}", e);
                None
            }
        }
    }

    /// Set a value in the cache with a TTL in seconds.
    pub async fn set(&self, key: &str, value: &str, ttl_seconds: u64) {
        match self.client.get_multiplexed_async_connection().await {
            Ok(mut conn) => {
                let result: Result<(), _> = conn.set_ex(key, value, ttl_seconds).await;
                if let Err(e) = result {
                    warn!("Redis cache_set error for key={}: {}", key, e);
                }
            }
            Err(e) => {
                warn!("Redis connection error in cache_set: {}", e);
            }
        }
    }

    /// Delete a key from the cache.
    pub async fn delete(&self, key: &str) {
        match self.client.get_multiplexed_async_connection().await {
            Ok(mut conn) => {
                let result: Result<(), _> = conn.del(key).await;
                if let Err(e) = result {
                    warn!("Redis cache_delete error for key={}: {}", key, e);
                }
            }
            Err(e) => {
                warn!("Redis connection error in cache_delete: {}", e);
            }
        }
    }
}
